SYNC-MUTATION.md

```diff
@@ -0,0 +1,335 @@
+    console.log(asyncVar.get()); // => 'value-1'
+  })
+
+asyncVar.set("value-2");
+new AsyncContext.Snapshot() // snapshot 2
+Promise.resolve()
+  .then(() => { // continuation 2
+    console.log(asyncVar.get()); // => 'value-2'
+  })
+```
+
+The value mapping is equivalent to:
+
+```
+⌌-----------⌍ snapshot 0
+|   'main'  |
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 1
+| 'value-1' |  <---- the continuation 1
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 2
+| 'value-2' |  <---- the continuation 2
+⌎-----------⌏
+```
+
+This trait is important with both `run` and `set` because mutations to
+`AsyncContext.Variable`s must not mutate prior `AsyncContext.Snapshot`s.
+
+> Note: this also applies to [`ContinuationVariable`][]
+
+### Decouple mutation with scopes
+
+To preserve the strong scope guarantees provided by `run`, an additional
+constraint can also be put to `set` to declare explicit scopes of mutation.
+
+A dedicated `AsyncContext.contextScope` can be decoupled with `run` to open a
+mutable scope with a series of `set` operations.
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.contextScope(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+});
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+`AsyncContext.contextScope` is basically a shortcut of
+`AsyncContext.Snapshot.run`:
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.Snapshot.wrap(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+})();
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+### Use cases
+
+One use case of `set` is that it allows more intuitive test framework
+integration (or similar frameworks that have prose style declarations,
+like middlewares).
+
+```js
+describe("asynct context", () => {
+  const ctx = new AsyncContext.Variable();
+
+  beforeEach((test) => {
+    ctx.set(1);
+  });
+
+  it('run in snapshot', () => {
+    // This function is run as a second paragraph of the test sequence.
+    assert.strictEqual(ctx.get(),1);
+  });
+});
+
+function testDriver() {
+  await AsyncContext.contextScope(async () => {
+    runBeforeEach();
+    await runTest();
+    runAfterEach();
+  });
+}
+```
+
+However, without proper test framework support, mutations in async `beforeEach`
+are still unintuitive, e.g. https://github.com/xunit/xunit/issues/1880.
+
+This will need a return-value API to feedback the final context snapshot to the
+next function paragraph.
```

Comment:
This sort of scenario is why I keep bringing up flow-through semantics. If context flows out of the return of that async function into subsequent execution then this interface is no longer needed.